modules/core/04-channel/keeper/upgrade.go

```diff
@@ -57,43 +61,222 @@ func (k Keeper) WriteUpgradeInitChannel(ctx sdk.Context, portID, channelID strin
 	emitChannelUpgradeInitEvent(ctx, portID, channelID, currentChannel, upgrade)
 }
 
-// upgradeTry
+// constructProposedUpgrade returns the proposed upgrade from the provided arguments.
+func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID string, fields types.UpgradeFields, upgradeTimeout types.Timeout) (types.Upgrade, error) {
+	seq, found := k.GetNextSequenceSend(ctx, portID, channelID)
+	if !found {
+		return types.Upgrade{}, types.ErrSequenceSendNotFound
+	}
+	return types.Upgrade{
+		Fields:             fields,
+		Timeout:            upgradeTimeout,
+		LatestSequenceSend: seq - 1,
+	}, nil
+}
+
+// ChanUpgradeTry is called by a module to accept the first step of a channel upgrade handshake initiated by
+// a module on another chain. If this function is successful, the proposed upgrade will be returned. If the upgrade fails, the upgrade sequence will still be incremented but an error will be returned.
 func (k Keeper) ChanUpgradeTry(
 	ctx sdk.Context,
 	portID,
 	channelID string,
 	proposedConnectionHops []string,
-	upgradeTimeout types.Timeout,
+	proposedUpgradeTimeout types.Timeout,
 	counterpartyProposedUpgrade types.Upgrade,
 	counterpartyUpgradeSequence uint64,
 	proofCounterpartyChannel,
-	proofCounterpartyUpgrade []byte,
+	proofUpgrade []byte,
```

Comment:
Are we missing two proof verification steps in this method or is the idea to add them later? Prove channel state and upgrade state?

Reply:
my understanding of the spec and `FlushUpgradeHandshake` API is that proof verification would be done there.